aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIO.java

```diff
@@ -393,4 +399,19 @@ public void close() {
       }
     }
   }
+
+  @SuppressWarnings("checkstyle:NoFinalizer")
+  @Override
+  protected void finalize() throws Throwable {
+    super.finalize();
+    if (!isResourceClosed.get()) {
+      if (null == createStack) {
+        close();
```

Comment:
This should call close either way, outside of the null check for `createStack`. The only behavior difference should be to only print the warning if the stack trace is present.